Issue5346 extract interface creates private classes

IvenBach opened the pull request:
Close #5346.

Still tracking down an issue where changing `ClassModule.Instancing` is not reflected properly in the View for the Extract Interface dialog. I'm stumped as to what's causing it. Documenting it here in case another duck can help me figure it out the root cause.

1. Have RD parse a class module with the code below while instancing is `Private`.
```
Option Explicit

Public Sub Bar()
    Debug.Print "bar"
End Sub

Public Sub Bazz()
    Debug.Print "Bazz"
End Sub

Public Sub Duk(value As String)
    Debug.Print value
End Sub
```

2. Change the instancing to `PublicNotCreatable` and perform a reparse. The Extract Interface dialog that shows is incorrect. The reason is a public implementing class mandates that the interface must also be public. Unfortunately as shown below the Private option is still enabled. Choose private and confirm with OK to create the private interface.
![Public implementing class has a private allowable interface which causes a compiler error](https://user-images.githubusercontent.com/26076874/72652333-3a2b6e80-393b-11ea-9c03-22f444e03313.png)

3. From the menu Debug>Compile Project and you have the case we're trying to eliminate. Delete the `Implements` keyword from the implementing class, then delete the newly created `IClass1` interface. This causes RD to reparse.

4. Go and extract another interface. Notice that now the Private option is greyed out, as it should have been originally. Confirm with OK to extract the interface and again Debug>Compile. This time everything is fine as both the implementing class and the interface are public.
![Private option greyed out with a public implementing class](https://user-images.githubusercontent.com/26076874/72652586-35b38580-393c-11ea-94bc-c8e3be7ac6ac.png)

MDoerner commented:
When you experienced the problem with the instancing not updating in RD, did you change the content of the module in any way or did you only changed the instancing?

If you only change the instancing, RD thinks that the module is unchanged because the code pane content has not changed. Consequently, the module is not reparse and the new information that only exists in the exported version of the code is not taken into account.

BZngr commented on Rubberduck.Core/UI/Converters/ClassInstancingToBooleanConverter.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    class ClassInstancingToBooleanConverter : IValueConverter
+    {
+        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
```
Maybe I just didn't spot it, but I don't see this converter class actually used in your xaml file.  Maybe its use was abandoned during development?  If it is used and retained, I'm thinking a name like `PublicNotCreatableInstancingToBoolConverter` would be more consistent with the other `<something>ToBoolConverter` converters since PublicNotCreatable <==> true.

BZngr commented:
When extracting an interface from a Public class, I think it would useful to hide both radio buttons as there is no choice for the user to make regarding the interface's exposure.  Or, both the Private and the Public (Not Creatable) radio button should be disabled.  Otherwise, an enabled Public (Not Creatable) button implies the user has the option to manipulate it.  Given the binary nature of this refactoring option, it seems that providing a single checkbox for the user to manipulate would also be sufficient.

IvenBach commented:
@MDoerner When I changed the instancing I tested it both ways. 1) Only changed the instancing, nothing else. 2) Changed instancing as well as a class members contents. Both ways resulted in a non-accurate instancing being shown in the extract interface dialog. The only way I've found to have instancing be accurate is by following the steps I wrote out above.

Vogel612 commented on Rubberduck.Refactorings/ExtractInterface/ExtractInterfaceRefactoring.cs:
```diff
@@ -129,17 +130,48 @@ private void AddInterfaceClass(Declaration implementingClass, string interfaceNa
+            try
+            {
+                var tempExportDirectory = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + System.IO.Path.DirectorySeparatorChar + "Rubberduck" + System.IO.Path.DirectorySeparatorChar;
+                var tempFile = interfaceComponent.ExportAsSourceFile(tempExportDirectory + interfaceComponent.Name);
```
could be replaced with `interfaceComponent.ExportToTempFile();`, removing the `tempExportDirectory` creation.

Vogel612 commented on Rubberduck.Refactorings/ExtractInterface/ExtractInterfaceRefactoring.cs:
```diff
@@ -129,17 +130,48 @@ private void AddInterfaceClass(Declaration implementingClass, string interfaceNa
+            }
+            catch (Exception)
+            {
+                throw;
```
I really don't like this catch block, because it does nothing useful here.

Vogel612 commented on Rubberduck.Refactorings/ExtractInterface/InterfaceMember.cs:
```diff
@@ -56,7 +56,7 @@ public InterfaceMember(Declaration member)
             Type = member.AsTypeName;
             
-            GetMethodType();
+            MemberType = GetMethodType(Member.Context);
```
should be `member.Context` to allow arbitrary reordering (even to before the assignment of `member` to `Member`)

Vogel612 commented on Rubberduck.Refactorings/InteractiveRefactoringBase.cs:
```diff
@@ -26,7 +26,8 @@ public abstract class InteractiveRefactoringBase<TPresenter, TModel> : Refactori
         {
-            Refactor(InitializeModel(target));
+            var model = InitializeModel(target);
+            Refactor(model);
```
I don't see the benefit of this local variable

Vogel612 commented on Rubberduck.Resources/RubberduckUI.resx:
```diff
@@ -1779,4 +1779,16 @@ Import aborted.</value>
+    <value>Public</value>
+  </data>
+  <data name="ExtractInterface_PublicInstancingMandatedByPublicClass" xml:space="preserve">
+    <value>'Public' will be chosen. For private instancing make implementing class private also.</value>
```
This needs a bit of reformulation, IMO. How about:

"Because the class the interface is extracted from is marked for public instancing, the interface must be public as well. If you want to extract a private interface, the class must be marked private before."

Vogel612 commented on Rubberduck.Core/UI/Refactorings/ExtractInterface/ExtractInterfaceView.xaml:
```diff
@@ -1,139 +1,154 @@
+            </GroupBox>
+
+            <GroupBox Header="{Resx ResxName=Rubberduck.Resources.RubberduckUI, Key=ExtractInterface_MembersGroupBox}"
+                      Grid.Row="2" Margin="5" Padding="5">
```
There's a full whitespace change here. Why?

MDoerner commented:
PR #5390 that just has been merged should have conflicts with this PR as it also adjusted the extract interface refactoring.

However, I would advise to wait until PR #5387 is merged, since it will conflict even more. However, it will also provide the necessary changes to make this PR work from the back-end side. (The change to the model has been taken from this PR.)

MDoerner commented:
If you merge next into this PR and take the side of next for the refactoring, the back-end for the UI enhancement should be there.

MDoerner commented:
I took the liberty to resolve the merge conflict in the resx file caused by gits inability to handle xml tags properly.

MDoerner commented:
Is there anything still to do in this PR or can the WIP label be removed?

IvenBach commented:
I consider this PR to add the correct wording complete.

That being said it still has the parsing issue originally described. The resolution for that is beyond the scope of this PR.
